File: 1036-rotting-oranges/rotting-oranges.py
from collections import deque
class Solution(object):
    def orangesRotting(self, grid):

        t=0
        m=len(grid)
        n=len(grid[0])
        for i in range(m):
            for j in range(n):
                if grid[i][j]==1:
                    t+=1
        if t==0:return 0
        def adja(m,n):
            L=[]
            for i in range(m):
                L.append([[] for _ in range(n)])
            return L

        for i in range(t):
            adj=adja(m,n)
            for j in range(m):
                for k in range(n):
                    if grid[j][k]==2:
                        if j+1<m:adj[j][k].append((j+1,k))
                        if j-1>=0:adj[j][k].append((j-1,k))
                        if k+1<n:adj[j][k].append((j,k+1))
                        if k-1>=0:adj[j][k].append((j,k-1))
            
            for j in range(m):
                for k in range(n):
                    if grid[j][k]==2:
                        for neighbour in adj[j][k]:
                            if grid[neighbour[0]][neighbour[1]]!=0:grid[neighbour[0]][neighbour[1]]=2

            ans=True
            for j in range(m):
                for k in range(n):
                    if grid[j][k]==1:
                        ans=False
                        break
                if not ans:
                    break
            if ans:return i+1
        return -1

File: 1036-rotting-oranges/rotting-oranges.cpp
#include <vector>
#include <queue>
using namespace std;

class Solution {
public:
    int orangesRotting(vector<vector<int>>& grid) {
        int m = grid.size(), n = grid[0].size();
        queue<pair<int, int>> q;
        int freshCount = 0, minutes = 0;

        // Count fresh oranges and enqueue the rotten ones
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                if (grid[i][j] == 2) {
                    q.push({i, j});
                } else if (grid[i][j] == 1) {
                    freshCount++;
                }
            }
        }

        // Directions for adjacent cells (up, down, left, right)
        vector<pair<int, int>> directions = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};

        // Perform BFS
        while (!q.empty() && freshCount > 0) {
            int size = q.size();
            for (int i = 0; i < size; ++i) {
                auto [x, y] = q.front();
                q.pop();
                for (auto [dx, dy] : directions) {
                    int nx = x + dx, ny = y + dy;
                    if (nx >= 0 && nx < m && ny >= 0 && ny < n && grid[nx][ny] == 1) {
                        grid[nx][ny] = 2;
                        q.push({nx, ny});
                        freshCount--;
                    }
                }
            }
            minutes++;
        }

        return freshCount == 0 ? minutes : -1;
    }
};
